### PR TITLE
lwAFTR uses new YANG-based configuration

### LINKED ISSUE #510: Migrate lwAFTR to Yang configuration
With the `yang` branch we will be migrating the lwaftr to YANG configuration, using the schema that currently lives at `src/apps/lwaftr/snabb-softwire.yang`.  Here is a random quick braindump of things to do.
- [x] Define a place for all yang schemas in the snabb source tree
- [x] Compile yang schemas into snabb binary
- [x] Allow yang schemas to be required by name via "require('yang.module.ietf-softwire') or something.  Perhaps that would load the text of the yang module, perhaps some compiled representation, probably the text tho which would then be parsed.
- [x] Tweak the yang module so that parsing produces a fresh data object
~~- [ ] Maintain a cache of loaded yang modules by name (?)~~ (trivial to do, will do if needed)
- [x] Add ability for a "compile" step that happens after validation.  Implement by require('yang.compile.ietf-softwire') which loads Lua code.  If no compile module, no problem.  For lwaftr this will compile "binding-entry" subtree to binding table.
- [x] Add binary serialize capability that writes out a validated, compiled configuration to a single file.  This is what we will use to cache compiled lwaftr configurations to speed up lwaftr startup, and as the configuration format passed between config and worker processes when workers need to do a full config reload.  A serialized configuration will embed the yang module, its version, and a version for the serializer, so that if anything is out of date, the cache won't be used.
- [x] Update snabb-softwire yang spec to match #519.
- [x] Add deserialize capability that will produce a YANG format configuration given a serialized file.  In this way also a running Snabb process will always be able to tell you the running in source format, or potentially a subtree of the configuration.
- [x] Add bridge to lwAFTR configuration from a data tree that has been compiled.  Trivial.
- [x] Switch configuration format of lwAFTR to compiled data tree

### PR DESCRIPTION
**Pull request by wingo**
This PR merges in the yang module to `lwaftr` and probably should cause a new release.  It changes the configuration format in an incompatible way, but there is the `snabb lwaftr migrate-configuration` command which we have used extensively as part of this migration and which can migrate the lwaftr to the new configuration.  Fixes #510.

### CODE REVIEW

**Comment by wingo** [comment]
An update here.  I didn't push to merge this because it removed the reload-configuration functionality.  That functionality is regained in #576, so once #576 is merged this PR will pick up those commits and thereby we can merge to lwaftr.

**Comment by wingo** [comment]
Ready to go, as it has the "get" and "load" commands that implement dump and reload, respectively!

**Comment by teknico** [comment]
As requested, here is the output of the `snabb lwaftr loadtest` command testing two instances of `snabb lwaftr run`, with the load described in #565, but in bump-in-the-wire mode, rather than on-a-stick as hinted there.

The first instance was running the current `lwaftr` branch, while the second was running this branch.

`lwaftr`:

```
Warming up at 5.000000 Gb/s for 2 seconds.
Applying 1.000000 Gbps of load.
  IPv4:
    TX 1426220 packets (0.285244 MPPS), 590771824 bytes (1.000002 Gbps)
    RX 1426220 packets (0.285244 MPPS), 647820624 bytes (1.091280 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 1306925 packets (0.261385 MPPS), 593634184 bytes (1.000001 Gbps)
    RX 1306925 packets (0.261385 MPPS), 541357184 bytes (0.916357 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 2.000000 Gbps of load.
  IPv4:
    TX 2852432 packets (0.570486 MPPS), 1181540528 bytes (1.999998 Gbps)
    RX 2852432 packets (0.570486 MPPS), 1295637808 bytes (2.182554 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 2613844 packets (0.522769 MPPS), 1187266224 bytes (1.999998 Gbps)
    RX 2613844 packets (0.522769 MPPS), 1082712464 bytes (1.832712 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 3.000000 Gbps of load.
  IPv4:
    TX 4278647 packets (0.855729 MPPS), 1772311056 bytes (2.999998 Gbps)
    RX 4278647 packets (0.855729 MPPS), 1943456936 bytes (3.273831 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 3920769 packets (0.784154 MPPS), 1780900408 bytes (2.999998 Gbps)
    RX 3920769 packets (0.784154 MPPS), 1624069648 bytes (2.749069 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 4.000000 Gbps of load.
  IPv4:
    TX 5704866 packets (1.140973 MPPS), 2363082272 bytes (3.999998 Gbps)
    RX 5704866 packets (1.140973 MPPS), 2591276912 bytes (4.365110 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 5227695 packets (1.045539 MPPS), 2374535240 bytes (4.000000 Gbps)
    RX 5227695 packets (1.045539 MPPS), 2165427440 bytes (3.665427 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 5.000000 Gbps of load.
  IPv4:
    TX 7131080 packets (1.426216 MPPS), 2953852568 bytes (4.999998 Gbps)
    RX 7131080 packets (1.426216 MPPS), 3239095768 bytes (5.456387 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 6534613 packets (1.306923 MPPS), 2968166632 bytes (4.999996 Gbps)
    RX 6534613 packets (1.306923 MPPS), 2706782112 bytes (4.581781 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 6.000000 Gbps of load.
  IPv4:
    TX 8557303 packets (1.711461 MPPS), 3544623720 bytes (5.999998 Gbps)
    RX 8557303 packets (1.711461 MPPS), 3886915840 bytes (6.547666 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 7841539 packets (1.568308 MPPS), 3561801464 bytes (5.999997 Gbps)
    RX 7841539 packets (1.568308 MPPS), 3248139904 bytes (5.498139 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 7.000000 Gbps of load.
  IPv4:
    TX 9983514 packets (1.996703 MPPS), 4135393936 bytes (6.999997 Gbps)
    RX 9249102 packets (1.849820 MPPS), 4220496710 bytes (7.107960 Gbps)
    Loss: 0 ingress drop + 734412 packets lost (7.356248%)
  IPv6:
    TX 9148464 packets (1.829693 MPPS), 4155435648 bytes (6.999998 Gbps)
    RX 9109554 packets (1.821911 MPPS), 3773407958 bytes (6.387260 Gbps)
    Loss: 0 ingress drop + 38910 packets lost (0.425317%)
Applying 8.000000 Gbps of load.
  IPv4:
    TX 11409734 packets (2.281947 MPPS), 4726166136 bytes (8.000000 Gbps)
    RX 9295342 packets (1.859068 MPPS), 4263202302 bytes (7.178065 Gbps)
    Loss: 0 ingress drop + 2114392 packets lost (18.531475%)
  IPv6:
    TX 10455390 packets (2.091078 MPPS), 4749070480 bytes (8.000000 Gbps)
    RX 9295334 packets (1.859067 MPPS), 3887154892 bytes (6.576389 Gbps)
    Loss: 0 ingress drop + 1160056 packets lost (11.095292%)
Applying 9.000000 Gbps of load.
  IPv4:
    TX 12835942 packets (2.567188 MPPS), 5316931656 bytes (8.999991 Gbps)
    RX 9277407 packets (1.855481 MPPS), 4259245606 bytes (7.171045 Gbps)
    Loss: 0 ingress drop + 3558535 packets lost (27.723209%)
  IPv6:
    TX 11762304 packets (2.352461 MPPS), 5342699280 bytes (8.999991 Gbps)
    RX 9277431 packets (1.855486 MPPS), 3885971004 bytes (6.573807 Gbps)
    Loss: 0 ingress drop + 2484873 packets lost (21.125734%)
Applying 10.000000 Gbps of load.
  IPv4:
    TX 14224372 packets (2.844874 MPPS), 5892050592 bytes (9.973497 Gbps)
    RX 9257275 packets (1.851455 MPPS), 4249940408 bytes (7.155384 Gbps)
    Loss: 451 ingress drop + 4966646 packets lost (34.919622%)
  IPv6:
    TX 13034720 packets (2.606944 MPPS), 5920659872 bytes (9.973589 Gbps)
    RX 9257233 packets (1.851447 MPPS), 3884521218 bytes (6.570712 Gbps)
    Loss: 464 ingress drop + 3777023 packets lost (28.980193%)
```

`yang`:

```
Warming up at 5.000000 Gb/s for 2 seconds.
Applying 1.000000 Gbps of load.
  IPv4:
    TX 1426218 packets (0.285244 MPPS), 590770608 bytes (1.000000 Gbps)
    RX 1426218 packets (0.285244 MPPS), 647819328 bytes (1.091278 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 1306925 packets (0.261385 MPPS), 593634184 bytes (1.000001 Gbps)
    RX 1306925 packets (0.261385 MPPS), 541357184 bytes (0.916357 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 2.000000 Gbps of load.
  IPv4:
    TX 2852433 packets (0.570487 MPPS), 1181541136 bytes (1.999999 Gbps)
    RX 2852433 packets (0.570487 MPPS), 1295638456 bytes (2.182555 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 2613845 packets (0.522769 MPPS), 1187267778 bytes (2.000000 Gbps)
    RX 2613845 packets (0.522769 MPPS), 1082713978 bytes (1.832714 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 3.000000 Gbps of load.
  IPv4:
    TX 4278646 packets (0.855729 MPPS), 1772310448 bytes (2.999997 Gbps)
    RX 4278646 packets (0.855729 MPPS), 1943456288 bytes (3.273830 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 3920768 packets (0.784154 MPPS), 1780898854 bytes (2.999996 Gbps)
    RX 3920768 packets (0.784154 MPPS), 1624068134 bytes (2.749067 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 4.000000 Gbps of load.
  IPv4:
    TX 5704866 packets (1.140973 MPPS), 2363082272 bytes (3.999998 Gbps)
    RX 5704866 packets (1.140973 MPPS), 2591276912 bytes (4.365110 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 5227694 packets (1.045539 MPPS), 2374534592 bytes (3.999999 Gbps)
    RX 5227694 packets (1.045539 MPPS), 2165426832 bytes (3.665426 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 5.000000 Gbps of load.
  IPv4:
    TX 7131086 packets (1.426217 MPPS), 2953854096 bytes (5.000000 Gbps)
    RX 7131086 packets (1.426217 MPPS), 3239097536 bytes (5.456390 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 6534620 packets (1.306924 MPPS), 2968169424 bytes (5.000000 Gbps)
    RX 6534620 packets (1.306924 MPPS), 2706784624 bytes (4.581785 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 6.000000 Gbps of load.
  IPv4:
    TX 8557309 packets (1.711462 MPPS), 3544627744 bytes (6.000005 Gbps)
    RX 8557309 packets (1.711462 MPPS), 3886920104 bytes (6.547673 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  IPv6:
    TX 7841549 packets (1.568310 MPPS), 3561806200 bytes (6.000005 Gbps)
    RX 7841549 packets (1.568310 MPPS), 3248144240 bytes (5.498146 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 7.000000 Gbps of load.
  IPv4:
    TX 9983511 packets (1.996702 MPPS), 4135392112 bytes (6.999994 Gbps)
    RX 9822085 packets (1.964417 MPPS), 4465621786 bytes (7.522163 Gbps)
    Loss: 0 ingress drop + 161426 packets lost (1.616926%)
  IPv6:
    TX 9148461 packets (1.829692 MPPS), 4155433704 bytes (6.999995 Gbps)
    RX 9124230 packets (1.824846 MPPS), 3779476486 bytes (6.397533 Gbps)
    Loss: 0 ingress drop + 24231 packets lost (0.264864%)
Applying 8.000000 Gbps of load.
  IPv4:
    TX 11409713 packets (2.281943 MPPS), 4726156480 bytes (7.999983 Gbps)
    RX 9638655 packets (1.927731 MPPS), 4415294260 bytes (7.434595 Gbps)
    Loss: 0 ingress drop + 1771058 packets lost (15.522371%)
  IPv6:
    TX 10455366 packets (2.091073 MPPS), 4749060160 bytes (7.999982 Gbps)
    RX 9638633 packets (1.927727 MPPS), 4020262762 bytes (6.802544 Gbps)
    Loss: 0 ingress drop + 816733 packets lost (7.811616%)
Applying 9.000000 Gbps of load.
  IPv4:
    TX 12835962 packets (2.567192 MPPS), 5316940704 bytes (9.000006 Gbps)
    RX 9643808 packets (1.928762 MPPS), 4423105470 bytes (7.447291 Gbps)
    Loss: 0 ingress drop + 3192154 packets lost (24.868833%)
  IPv6:
    TX 11762324 packets (2.352465 MPPS), 5342708752 bytes (9.000007 Gbps)
    RX 9643775 packets (1.928755 MPPS), 4033415292 bytes (6.823785 Gbps)
    Loss: 0 ingress drop + 2118549 packets lost (18.011313%)
Applying 10.000000 Gbps of load.
  IPv4:
    TX 14225712 packets (2.845142 MPPS), 5892607046 bytes (9.974439 Gbps)
    RX 9628267 packets (1.925653 MPPS), 4419969124 bytes (7.441676 Gbps)
    Loss: 810 ingress drop + 4596635 packets lost (32.317855%)
  IPv6:
    TX 13036229 packets (2.607246 MPPS), 5921344712 bytes (9.974743 Gbps)
    RX 9628536 packets (1.925707 MPPS), 4040564864 bytes (6.834640 Gbps)
    Loss: 504 ingress drop + 3407189 packets lost (26.140174%)
```

**Comment by wingo** [comment]
Thank you @teknico for those numbers!  I recall for people not following that this isn't our 550-byte workload but rather an IMIX workload as specified in that mentioned bug.  It has a lower average packet size.  I find it interesting that we don't hit 2MPPS FDX with this IMIX on `lwaftr` but that's neither here nor there; the interesting bit is that even with the binding table loaded from the `ctable` that is automatically generated by the yang code, our perf improves.  The improvement is surely due to the topological-sort code that this patchset pulls in.  Anyway it's enough to know that we're not regressing (and actually we appear to be progressing) so I think we're good for merging this one in tomorrow; wdyt @kbara?

File: src/apps/config/action_codec.lua [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+end
+
+local lower_case = "abcdefghijklmnopqrstuvwxyz"
+local upper_case = lower_case:upper()
+local extra = "0123456789_-"
+local alphabet = table.concat({lower_case, upper_case, extra})
+assert(#alphabet == 64)
+local function random_file_name()
+   local f = io.open('/dev/urandom', 'rb')
+   -- 22 bytes, but we only use 2^6=64 bits from each byte, so total of
```
Comment (kbara) [inline src/apps/config/action_codec.lua:92]: Calling that 64 bits is wrong, though the intention is clear; it's 6 bits. [Not a blocker, no change needed before release]

File: src/lib/ipsec/README.md [hunk 2]
```diff
@@ -13,12 +13,25 @@ UDP, L2TPv3) and also encrypts the contents of the inner protocol
...
 
-Anti-replay protection as well as recovery from synchronization loss due to
-excessive packet loss are *not* implemented.
+    DIAGRAM: ESP-Transport
+         BEFORE ENCAPSULATION
+    +-----------+-------------+------------+
+    |orig Ether‑| orig IP Hdr |            |
+    |net Hdr    |(any options)|  Payload   |
+    +-----------+-------------+------------+
+
```
Comment (kbara) [inline src/lib/ipsec/README.md:22]: Not a blocker, no change needed before release, but whitespace errors.

**Comment by kbara** [comment]
LGTM

**Pull request merged**